core/src/batch/replace.rs

```diff
@@ -0,0 +1,69 @@
+use super::{Batch, Disposition};
+use crate::packets::Packet;
+use crate::Result;
+
+/// A batch that replaces each packet of the batch with another packet.
+///
+/// The original packet is dropped from the batch with the new packet in its
+/// place. On error, the packet is `aborted` and will short-circuit the
+/// remainder of the pipeline.
+pub struct Replace<B: Batch, T: Packet, F>
+where
+    F: FnMut(&B::Item) -> Result<T>,
+{
+    batch: B,
+    f: F,
+    slot: Option<B::Item>,
+}
+
+impl<B: Batch, T: Packet, F> Replace<B, T, F>
+where
+    F: FnMut(&B::Item) -> Result<T>,
+{
+    #[inline]
+    pub fn new(batch: B, f: F) -> Self {
+        Replace {
+            batch,
+            f,
+            slot: None,
+        }
+    }
+}
+
+impl<B: Batch, T: Packet, F> Batch for Replace<B, T, F>
+where
+    F: FnMut(&B::Item) -> Result<T>,
+{
+    type Item = T;
+
+    #[inline]
+    fn replenish(&mut self) {
+        self.batch.replenish();
+    }
+
+    #[inline]
+    fn next(&mut self) -> Option<Disposition<Self::Item>> {
+        // internally the replace combinator will add a new packet to the
+        // batch and mark the original as dropped. each packet in the
+        // batch could become 2.
+        if let Some(pkt) = self.slot.take() {
```

Comment:
more of a q, thoughts if this being some inefficient? I don't think it's a big deal, but worth tasking where it could hurt us compared to another approach?

Reply:
yeah it grows the iter to 2x. but should be better than the out of band queuing approach we did before.

Reply:
yeah. I may put that as a note, but nonetheless.